packages/aggregator/src/__tests__/analytics/aggregateAnalytics/aggregations/getFinalValuePerPeriod.test.js

```diff
@@ -0,0 +1,128 @@
+import { getPeriodsInRange, getCurrentPeriod, convertToPeriod } from '@tupaia/utils';
+import { getFinalValuePerPeriod } from '../../../../analytics/aggregateAnalytics/aggregations';
+
+describe('getFinalValuePerPeriod()', () => {
+  describe('getDistinctValues()', () => {
+    it('change period to WEEK', () => {
+      const testAnalytics = [
+        { dataElement: 'element1', organisationUnit: 'org1', period: '20210101', value: 1 },
+        { dataElement: 'element1', organisationUnit: 'org2', period: '20210102', value: 2 },
+        { dataElement: 'element1', organisationUnit: 'org1', period: '20210108', value: 3 },
+        { dataElement: 'element1', organisationUnit: 'org2', period: '20210109', value: 3 },
+      ];
+      const aggregationPeriod = 'WEEK';
+      expect(getFinalValuePerPeriod(testAnalytics, null, aggregationPeriod)).toStrictEqual([
+        { dataElement: 'element1', organisationUnit: 'org1', period: '2020W53', value: 1 },
+        { dataElement: 'element1', organisationUnit: 'org1', period: '2021W01', value: 3 },
+        { dataElement: 'element1', organisationUnit: 'org2', period: '2020W53', value: 2 },
+        { dataElement: 'element1', organisationUnit: 'org2', period: '2021W01', value: 3 },
+      ]);
+    });
+
+    it('get final value', () => {
+      const testAnalytics = [
+        { dataElement: 'element1', organisationUnit: 'org1', period: '20210101', value: 1 },
+        { dataElement: 'element1', organisationUnit: 'org1', period: '20210102', value: 2 },
+        { dataElement: 'element1', organisationUnit: 'org2', period: '20210108', value: 1 },
+        { dataElement: 'element1', organisationUnit: 'org2', period: '20210109', value: 3 },
+      ];
+      const aggregationPeriod = 'WEEK';
+      expect(getFinalValuePerPeriod(testAnalytics, null, aggregationPeriod)).toStrictEqual([
+        { dataElement: 'element1', organisationUnit: 'org1', period: '2020W53', value: 2 },
+        { dataElement: 'element1', organisationUnit: 'org2', period: '2021W01', value: 3 },
+      ]);
+    });
+  });
+
+  describe('getFilledValues()', () => {
+    const getExpectedArray = (startDate, aggregationPeriod, defaultVariables) => {
+      const startPeriod = convertToPeriod(startDate, aggregationPeriod);
+      const endPeriod = getCurrentPeriod(aggregationPeriod);
+      const periods = getPeriodsInRange(startPeriod, endPeriod);
+      return periods.map(period => ({
+        ...defaultVariables,
+        period,
+      }));
+    };
+
+    it('Filling continuous values', () => {
+      const defaultVariables = { dataElement: 'element1', organisationUnit: 'org1', value: 1 };
+
+      const startDate = '20210101';
+      const aggregationPeriod = 'WEEK';
+      const testAnalytics = [{ ...defaultVariables, period: startDate }];
```

Comment:
Probably worth having more than just 1 analytic here, to properly test that `previous` logic is working correctly

Reply:
Good idea! Will add that to the test.